sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ComponentLoader.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.autoconfigure.internal;
+
+import io.opentelemetry.sdk.autoconfigure.spi.Ordered;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A loader for components that are discovered via SPI.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public interface ComponentLoader {
+  /**
+   * Load implementations of an SPI.
+   *
+   * @param spiClass the SPI class
+   * @param <T> the SPI type
+   * @return iterable of SPI implementations
+   */
+  <T> Iterable<T> load(Class<T> spiClass);
+
+  /**
+   * Load implementations of an ordered SPI (i.e. implements {@link Ordered}).
+   *
+   * @param spiClass the SPI class
+   * @param <T> the SPI type
+   * @return list of SPI implementations, in order
+   */
+  default <T extends Ordered> List<T> loadOrdered(Class<T> spiClass) {
+    return StreamSupport.stream(load(spiClass).spliterator(), false)
+        .sorted(Comparator.comparing(Ordered::order))
+        .collect(Collectors.toList());
+  }
```

Comment:
do we need this method? (can inline as private method in `SpiHelper`?)

Reply:
the method is meant for overriding. 
In spring, the spring beans are given a higher priority - https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10453/files#diff-8c1885af89564bfbd461edbbf02fe19e23e40a2e54f6e264d41d786b760e5d26R40-R53.

I'm still unsure if that's what users actually expect though.

It affects
- AutoConfigurationCustomizerProvider
- ResourceProvider

Reply:
can't find a really compelling reason - I'll remove it.